harmonizer/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+use deno_core::Op;
+use deno_core::{json_op_sync, JsRuntime};
+use serde::{Deserialize, Serialize};
+use std::sync::mpsc::channel;
+use std::{fmt::Display, io::Write};
+use thiserror::Error;
+
+#[derive(Serialize)]
+// When serialized, we'll be putting this into JavaScript expecting camelCase.
+#[serde(rename_all = "camelCase")]
+pub struct ServiceDefinition {
+    pub name: String,
+    pub url: String,
+    pub type_defs: String,
+}
+
+impl ServiceDefinition {
+    pub fn new<N: Into<String>, U: Into<String>, D: Into<String>>(
+        name: N,
+        url: U,
+        type_defs: D,
+    ) -> ServiceDefinition {
+        ServiceDefinition {
+            name: name.into(),
+            url: url.into(),
+            type_defs: type_defs.into(),
+        }
+    }
+}
+
+pub type ServiceList = Vec<ServiceDefinition>;
+
+#[derive(Debug, Error, Serialize, Deserialize, PartialEq)]
+pub struct CompositionError {
+    pub message: Option<String>,
+    pub extensions: Option<CompositionErrorExtensions>,
+}
+
+impl Display for CompositionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(msg) = &self.message {
+            f.write_fmt(format_args!("{code}: {msg}", code = self.code(), msg = msg))
+        } else {
+            f.write_str(self.code())
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct CompositionErrorExtensions {
+    pub code: String,
+}
+
+impl CompositionError {
+    pub fn code(&self) -> &str {
+        if let Some(ref ext) = self.extensions {
+            ext.code.as_str()
+        } else {
+            "UNKNOWN"
+        }
+    }
+}
+
+pub fn harmonize(service_list: ServiceList) -> Result<String, Vec<CompositionError>> {
+    // Initialize a runtime instance
+    let mut runtime = JsRuntime::new(Default::default());
+
+    // We'll use this channel to get the results
+    let (tx, rx) = channel();
+
+    // The first thing we do is define an op so we can print data to STDOUT,
+    // because by default the JavaScript console functions are just stubs (they
+    // don't do anything).
+
+    // Register the op for outputting bytes to stdout. It can be invoked with
+    // Deno.core.dispatch and the id this method returns or
+    // Deno.core.dispatchByName and the name provided.
+    runtime.register_op(
+        "op_print",
+        // The op_fn callback takes a state object OpState
+        // and a vector of ZeroCopyBuf's, which are mutable references
+        // to ArrayBuffer's in JavaScript.
+        |_state, zero_copy| {
+            let mut out = std::io::stdout();
+
+            // Write the contents of every buffer to stdout
+            for buf in zero_copy {
+                out.write_all(&buf).unwrap();
```

Comment:
If `unwrap` is necessary, I'd use `expect` instead.

Reply:
Fixed in 3d9967d1ccb6323f877052ef4be433e682844220 (many fixes!)